docs/developers_guide/FlinkComponents.md

```diff
@@ -1,22 +1,86 @@
+is more complicated since the following things has to be provided:
+- a Flink [SourceFunction](https://nightlies.apache.org/flink/flink-docs-stable/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.html)
+- Flink [type information](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/datastream/fault-tolerance/serialization/types_serialization/#flinks-typeinformation-class)
+  for serializing/deserializing the emitted `#input`
+- a [timestamp watermark handler](https://github.com/TouK/nussknacker/blob/staging/engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/timestampwatermark/TimestampWatermarkHandler.scala)
+  so that events are correctly processed downstream, for example to avoid (or force!) dropping late events by aggregates. Read more about
+  [notion of time](../scenarios_authoring/Intro.md#streaming-flink-only-notion-of-time)
+  and [watermarks](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/datastream/event-time/generating_watermarks/)
+- (optionally) generating test data support (trait `FlinkSourceTestSupport`) to ease [scenarios authoring](../scenarios_authoring/TestingAndDebugging.md)
+- (optionally) if the source returns a record having generic type and its fields and methods cannot be discovered using Java reflection,
+  [ReturningType](https://github.com/TouK/nussknacker/blob/staging/components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/ReturningType.scala)
+  can be implemented to provide detailed typing info for the emitted `#input`. See here for more details about [types](./Basics.md#types).
+  Please note returning type (and much more, like dynamic parameters) can be also specified with [GenericNodeTransformation](./Components.md#genericnodetransformation)
+- (optionally) custom [context initializer](https://github.com/TouK/nussknacker/blob/staging/components-api/src/main/scala/pl/touk/nussknacker/engine/api/process/ContextInitializer.scala)
+  to emit more variables than `#input`. For example built-in Kafka sources emits `#inputMeta` variable with Kafka record metadata like: partition, topic, offset, etc.
+  The other example could be a file source that emits current line number as a new variable along with the content (as `#input` variable)
+
+Nussknacker also provides a more generic [FlinkSource](https://github.com/TouK/nussknacker/blob/staging/engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala)
+for implementing sources. The difference is instead of implementing a Flink `SourceFunction`, arbitrary `DataStream[Context]`
+can be returned, however you have to remember to assign timestamps, watermarks and initialize the context.
+
+The above `BasicFlinkSource` and `FlinkSource` are **only** implementation. They have to be returned by [SourceFactory](https://github.com/TouK/nussknacker/blob/staging/components-api/src/main/scala/pl/touk/nussknacker/engine/api/process/Source.scala)
```

Comment:
I think this part is unclear. Maybe sth along the lines:
When using Flink engine, all sources returned by `SourceFactory` **have** to implement 
`FlinkSource` (or its subtrait `BasicFlinkSource`) 

(same for sinks)